cranelift-codegen/meta/src/cdsl/instructions.rs

```diff
@@ -408,36 +410,126 @@ impl ValueTypeOrAny {
     }
 }
 
+/// The number of bits in the vector
+type VectorBitWidth = u64;
+
+/// An instruction paramater used for binding instructions to specific types or values
+pub enum InstructionParameter {
+    AnyType,
+    LaneType(LaneType),
+    VectorType(LaneType, VectorBitWidth),
+    ReferenceType(ReferenceType),
+    ImmediateValue(Immediate),
+}
+
+impl From<shared::types::Int> for InstructionParameter {
+    fn from(ty: shared::types::Int) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<shared::types::Bool> for InstructionParameter {
+    fn from(ty: shared::types::Bool) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<shared::types::Float> for InstructionParameter {
+    fn from(ty: shared::types::Float) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<cdsl::types::LaneType> for InstructionParameter {
+    fn from(ty: cdsl::types::LaneType) -> Self {
+        InstructionParameter::LaneType(ty.into())
+    }
+}
+
+impl From<shared::types::Reference> for InstructionParameter {
+    fn from(ty: shared::types::Reference) -> Self {
+        InstructionParameter::ReferenceType(ty.into())
+    }
+}
+
+#[derive(Clone)]
+pub enum Immediate {
+    IntCC(u8),
+    FloatCC(u8),
+    UInt8(u8),
+    UInt128(u128),
+}
```

Comment:
If we are fine with doing string comparisons at compile-time, this could be removed and a string could be passed to `bind` as the immediate value.